Packages/com.unity.inputsystem/InputSystem/Plugins/OnScreen/OnScreenControl.cs

```diff
@@ -25,18 +27,140 @@ namespace UnityEngine.Experimental.Input.Plugins.OnScreen
     /// </remarks>
     public abstract class OnScreenControl : MonoBehaviour
     {
-        [SerializeField] private string m_ControlPath;
+        private struct OnScreenDeviceEventInfo
+        {
+            public InputEventPtr eventPtr;
+            public NativeArray<byte> buffer;
+            public InputDevice device;
+        }
+
+        private static InlinedArray<OnScreenDeviceEventInfo> s_DeviceEventInfoArray = new InlinedArray<OnScreenDeviceEventInfo>();
+
+        // We will have N devices mapped to X number of OnScreenControls
+        // Need to keep track and reference count how many controls are sharing
+        // the same device, so that when the last control is ready to be destroyed
+        // it can clean up the memory and remove the device from the input system.
+        private static List<int> s_RegisteredOnScreenControls = new List<int>();
```

Comment:
This is more convoluted than it needs to be. Simply have a `public OnScreenControl firstControl` on `OnScreenDeviceEventInfo` and a `private OnScreenControl m_NextControlOnDevice;` in `OnScreenControl`. In `OnEnable` look up / create entry and add to list. In `OnDisable`, remove from list and nuke device if last OSC disappeared from entry.